state_tracker.go

```diff
@@ -17,126 +17,43 @@ import (
-// structs):
-//
-//  BinlogStateTracker       BinlogStateTracker
-// IterationStateTracker   IterationStateTracker
-//          |                        |
-//          v                        v
-//   CopyStateTracker       VerifierStateTracker
-//          |                        |
-//          +-----------+------------+
-//                      |
-//                      v
-//                 StateTracker
-//                      |
-//                      v
-//                    Ferry
-//
-// The same relationship exists for the serializable states.
-
-type BinlogSerializableState struct {
-	LastProcessedBinlogPosition mysql.Position
-}
-
-type IterationSerializableState struct {
-	LastSuccessfulPrimaryKeys map[string]uint64
-	CompletedTables           map[string]bool
-}
+// The states stored in the state tracker can be copied into a
+// serialization-friendly struct (`SerializableState`), which can then be
+// dumped using something like JSON. Assuming the rest of Ghostferry used the
+// API of the state tracker correctlym this can be done at any point during the
+// Ghostferry run and the resulting state can be resumed from without data
+// loss.  The same `SerializableState` is used as an input to `Ferry`, which
+// will instruct the `Ferry` to resume a previously interrupted run.
 
-type CopySerializableState struct {
-	*BinlogSerializableState
-	*IterationSerializableState
-}
-
-type VerificationSerializableState struct {
-	*BinlogSerializableState
-	*IterationSerializableState
-
-	// This is not efficient because we have to build this map of a different
-	// type from the original ReverifyStore struct.
-	//
-	// TODO: address this inefficiency later.
-	ReverifyStore map[string][]uint64
-}
-
-const (
-	StageCopy         = "COPY"
-	StageVerification = "VERIFICATION"
-)
-
-// This is the struct that is dumped by Ghostferry when it is interrupted. It
-// is the same struct that is given to Ghostferry when it is resumed.
 type SerializableState struct {
 	GhostferryVersion         string
 	LastKnownTableSchemaCache TableSchemaCache
 
-	CurrentStage  string
-	CopyStage     *CopySerializableState
-	VerifierStage *VerificationSerializableState
+	LastSuccessfulPrimaryKeys                 map[string]uint64
+	CompletedTables                           map[string]bool
+	LastWrittenBinlogPosition                 mysql.Position
+	LastStoredBinlogPositionForInlineVerifier mysql.Position
 }
 
-// The binlog writer and the verify binlog positions are different because the
-// binlog writer is buffered in a background go routine. Its position can race
-// with respect to the verifier binlog position. The minimum position between
-// the two are always safe to resume from.
 func (s *SerializableState) MinBinlogPosition() mysql.Position {
-	if s.VerifierStage == nil {
-		return s.CopyStage.LastProcessedBinlogPosition
+	nilPosition := mysql.Position{}
+	if s.LastWrittenBinlogPosition == nilPosition {
+		return s.LastStoredBinlogPositionForInlineVerifier
```

Comment:
If there was no `LastWrittenBinlogPosition` (nothing was written by the binlog writer), how could we have a `LastStoredBinlogPositionForInlineVerifier`?

Reply:
Both the `LastWrittenBinlogPosition` and the `LastStoredBinlogPositionForInlineVerifier` are initialized during `Ferry.Start`, right after the binlog streamer is connected. This is done to address a bug where if the application database got not writes to it while Ghostferry is running, Ghostferry could not resume.